modules/cnf-topology-aware-lifecycle-manager-precache-user-spec-images.adoc

```diff
@@ -0,0 +1,165 @@
++
+[source,yaml]
+----
+apiVersion: ran.openshift.io/v1alpha1
+kind: PreCachingConfig
+metadata:
+  name: exampleconfig
+  namespace: default <1>
+spec:
+  overrides:
+    platformImage: quay.io/openshift-release-dev/ocp-release@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
+    operatorsIndexes:
+      - registry.example.com:5000/custom-redhat-operators:1.0.0
+    operatorsPackagesAndChannels:
+      - local-storage-operator: stable
+      - ptp-operator: stable
+      - sriov-network-operator: stable
+  spaceRequired: 30Gi
+  excludePrecachePatterns:
+    - aws
+    - vsphere
+  additionalImages:
+    - quay.io/exampleconfig/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
+    - quay.io/exampleconfig/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
+    - quay.io/exampleconfig/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
+----
+<1> The `namespace` must be accessible to the hub cluster.
+
+. Create a `ClusterGroupUpgrade` CR with the `preCaching` field set to `true` and specify the `PreCachingConfig` CR created in the previous step:
++
+[source,yaml]
+----
+apiVersion: ran.openshift.io/v1alpha1
+kind: ClusterGroupUpgrade
+metadata:
+  name: cgu
+  namespace: default
+spec:
+  clusters:
+  - sno1
+  - sno2
+  preCaching: true
+  preCachingConfigRef:
+  - name: exampleconfig
+    namespace: default
+  remediationStrategy:
+    timeout: 240
+----
+
+. {cgu-operator} verifies the `ClusterGroupUpgrade` CR.
+If the verification is successful, the pre-caching reconciliation starts.
++
+[NOTE]
+====
+All sites are pre-cached concurrently.
+====
+// What happens if the verification fails? Any common mistakes? Troubleshooting steps?
+
+.Verification
+// This is all automatic. But what if something goes wrong? Is this all automatically printed or do you have to request?
+
+. The pre-caching configurations are validated by checking if the managed policies exist.
+A valid configuration results in the following status:
++
+.Example of validated PreCachingConfig CR
+[source,terminal]
+----
+Type:    "PrecacheSpecValid",
+Status:  True,
+Reason:  "PrecacheSpecIsWellFormed",
+Message: "Precaching spec is valid and consistent"
+----
+
+. With a valid configuration, the pre-caching job starts with the following status update:
++
+.Example of pre-caching job in progress
+[source,terminal]
+----
+Type:    "PrecachingSuceeded",
+Status:  False,
+Reason:  "InProgress",
+Message: "Precaching is required and not done"
+----
+
+. The pre-caching job is successfully completed with the following status update:
++
+.Example of completed pre-caching job
+[source,terminal]
+----
+Type:    "PrecachingSuceeded",
+Status:  True,
+Reason:  "PrecachingCompleted",
+Message: "Precaching is completed for all clusters"
+----
+
+From this point, you can continue with the {cgu-operator} pre-caching workflow.
```

Comment:
1) User creates a CGU CR with the preCaching field set to true and prepares a pre-caching config CR with the list of additional images to be pre-cached. The user may also specify the space required for the additional images in order to avoid triggering kubelet to run GC. 
**Note** that the CR should be created concurrently with the CGU or should exist prior to the creation of it. Furthermore, the pre-caching config CR should be created in a namespace accessible to the hub cluster.

2) The CGU CR is validated (a function verifies if the clusters specified in the CR are part of a ManagedCluster object).

3) If the validation succeeds, the pre-caching reconciliation commences. Note that all the sites are pre-cached concurrently (consistent with existing behavior). From here onwards. the same workflow exists as that detailed in the referenced url.

Reply:
The CGU should be initially applied with the setting `preCaching: true`. Thereafter, with consequent successful validations, the pre-caching job will naturally progress to the subsequent stage in the workflow (i.e. no user intervention required). No additional changes to the CGU are required.